README.md
# pyeasyeda

[![ci-cd](https://github.com/UBC-MDS/pyeasyeda/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/pyeasyeda/actions/workflows/ci-cd.yml) [![codecov](https://codecov.io/gh/UBC-MDS/pyeasyeda/branch/main/graph/badge.svg?token=vaOyqFqkor)](https://codecov.io/gh/UBC-MDS/pyeasyeda)

Since exploratory data analysis is an imperative part of every analysis, this package aims at providing efficient data scrubbing and visualization tools to perform preliminary EDA on raw data. The package can be leveraged to clean the dataset and visualize relationships between features to generate insightful trends.

## Functions

-   `clean_up` - This function takes in a pandas dataframe object and performs initial steps of EDA on unstructured data. It returns a clean dataset by removing null values and identifying potential outliers in numeric variables based on a defined threshold.

-   `birds_eye_view` - This function takes in a pandas dataframe object and visualizes the distributions of variables in the form of histograms and density plots. It also generates a correlation heatmap for numeric variables to study their relationships.

-   `close_up` - This function accepts a pandas dataframe object creates a scatterplot of the variable(s) most strongly correlated with the dependent variable. The plot also produces a trend line to model the correlation between the variables.

-   `summary_suggestions` - This function takes in a pandas dataframe object and outputs a table of summary statistics for numeric and categorical variables and a table for percentage of unique values in the categorical variables.

Other packages that offer similar functionality are:
- [datascience_eda](https://github.com/UBC-MDS/datascience_eda)
- [QuickDA](https://github.com/sid-the-coder/QuickDA)
- [easy-data-analysis](https://github.com/jschnab/easy-data-analysis)

## Installation

```bash
$ pip install pyeasyeda
```

## Usage
After installing the package through the command above, please run the following commands from the root of the project repo as a quick demo. 
```
python
import pandas as pd
from pyeasyeda.clean_up import clean_up
from pyeasyeda.birds_eye_view import birds_eye_view
from pyeasyeda.close_up import close_up
from pyeasyeda.summary_suggestions import summary_suggestions
df = pd.read_csv("tests/data/penguins_test.csv")
clean_up(df)
plots = birds_eye_view(df)
close_up(df, 1)
summary_suggestions(df)

```
Please check our official documentation for the example usage of the package at [pyeasyeda/example](https://pyeasyeda.readthedocs.io/en/latest/example.html) on Read the Docs.

## Documentation
The official documentation is hosted at [pyeasyeda](https://pyeasyeda.readthedocs.io/en/latest/) on Read the Docs.

## Contributors
This python package was developed by James Kim, Kristin Bunyan, Luming Yang and Sukhleen Kaur. The team is from the Master of Data Science program at the University of the British Columbia.

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`pyeasyeda` was created by James Kim, Kristin Banyan, Luming Yang and Sukhleen Kaur. It is licensed under the terms of the MIT license.

## Credits

`pyeasyeda` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.2.8 (2022-02-04)


## v0.2.7 (2022-01-28)


## v0.2.6 (2022-01-28)


## v0.2.5 (2022-01-28)


## v0.2.4 (2022-01-27)


## v0.2.3 (2022-01-27)


## v0.2.2 (2022-01-27)


## v0.2.1 (2022-01-27)


## v0.2.0 (2022-01-27)
### Feature
* Close_up ([`24329e8`](https://github.com/UBC-MDS/pyeasyeda/commit/24329e8bebf3286c29900cc6442aee0c3d881913))

### Documentation
* Updated example demonstration ([`437426f`](https://github.com/UBC-MDS/pyeasyeda/commit/437426f865cebe53a7cac7e35bf2cdd7f23dd9f2))
* Update close_up descriptions ([`881197b`](https://github.com/UBC-MDS/pyeasyeda/commit/881197b4db42d4eaea4dfa8ebb04442913699e98))
* Update docstrings for close_up function ([`2c1d5c7`](https://github.com/UBC-MDS/pyeasyeda/commit/2c1d5c700bd44f7b6211ab8e06d3e8d5d05b3f13))

## v0.1.0 (11/01/2022)

- First release of `pyeasyeda`!

tests/test_clean_up.py
import pandas as pd
import numpy as np
import pytest
from pytest import raises
from scipy import stats
import io
import sys
from pyeasyeda.clean_up import clean_up


def test_clean_up():
    """tests the new dataset and output outliers of clean_up from a toy dataset."""
    
    # Load toy dataset
    data = {
        "neighbourhood": ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"],
        "price": [100, 120, 150, 90, 150000, np.NaN, 200, 300, 500, 50, 300, 100, 200],
        "number_of_reviews": [600, 1, 27000000, 1, 1, 2, 1, 2, 0, np.NaN, 2, 100, 500]
    }

    df = pd.DataFrame(data)

    clean = pd.DataFrame(
        {
        "neighbourhood": ["A", "B", "C", "D", "E", "G", "H", "I", "K", "L", "M"],
        "price": [100.0, 120.0, 150.0, 90.0, 150000.0, 200.0, 300.0, 500.0, 300.0, 100.0, 200.0],
        "number_of_reviews": [600.0, 1.0, 27000000.0, 1.0, 1.0, 1.0, 2.0, 0.0, 2.0, 100.0, 500.0]
        }
    )

    # Test the returned dataframe
    assert pd.DataFrame.equals(
        clean_up(df), clean
    ), "The returned dataframe using clean_up is not correct"
    
    # Create StringIO object
    capturedOutput = io.StringIO()
    
    # Redirect stdout
    sys.stdout = capturedOutput
    
    # Call clean_up on toy dataset
    clean_up(df)
    
    # Reset redirect
    sys.stdout = sys.__stdout__
    
    # Store print output in statement as a string
    statement = str(capturedOutput.getvalue())
    
    # Check if the statement contains the correct outliers
    assert (
        '2.7e+07' in statement
    ), "Outlier should contain 2.7e+07"
    assert (
        '150000' in statement
    ), "Outlier should contain 150000"

def test_clean_up_error():
    """Check TypeError and ValueError raised when inputs are not appropriate."""

   # Load toy dataset
    data = {
        "neighbourhood": ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M"],
        "price": [100, 120, 150, 90, 150000, np.NaN, 200, 300, 500, 50, 300, 100, 200],
        "number_of_reviews": [600, 1, 27000000, 1, 1, 2, 1, 2, 0, np.NaN, 2, 100, 500]
    }

    df = pd.DataFrame(data)

    # Tests if the input is not dataframe
    with raises (TypeError):
        clean_up("not dataframe")


tests/test_close_up.py
import pandas as pd
import numpy as np
import altair as alt
from pyeasyeda.close_up import close_up
import pytest


def test_close_up():
    """Tests close_up from a toy dataset."""

    # create toy_data
    toy_data = {
        "income": [5, 8, 10, 12, 17, 19],
        "house_size": [700, 600, 900, 1000, 1200, 1500],
        "views": ["mountain", "river", "sea", "mountain", "urban", "forest"],
        "price": [65, 50, 80, 98.5, 112, 133],
        "doctor_visits": [6, 8, 4, 5, 3, 2],
    }
    df = pd.DataFrame(toy_data)

    # unit tests - returned altair chart object
    fig = close_up(df, 1)
    assert isinstance(
        fig, alt.vegalite.v4.api.LayerChart
    ), "n=1 should return a LayerChart"
    actual = str(type(close_up(df, 1)))

    fig = close_up(df, 4)
    assert isinstance(
        fig, alt.vegalite.v4.api.VConcatChart
    ), "n>1 should return a VConcatChart"

    assert len(fig.vconcat) == 4, "Incorrect number of subplots!"

    assert (fig.data == df).sum().sum(
    ) == 30, "Data used in plot is different."


def test_close_up_error():
    """Check TypeError and ValueError raised when inputs are not appropriate."""
    with pytest.raises(TypeError):
        numpy_array = np.ones((3, 3))
        close_up(numpy_array)

    # create toy_data
    toy_data = {
        "income": [5, 8, 10, 12, 17, 19],
        "house_size": [700, 600, 900, 1000, 1200, 1500],
        "views": ["mountain", "river", "sea", "mountain", "urban", "forest"],
        "price": [65, 50, 80, 98.5, 112, 133],
        "doctor_visits": [6, 8, 4, 5, 3, 2],
    }
    df = pd.DataFrame(toy_data)

    with pytest.raises(TypeError):
        float_num = 2.3
        close_up(df, n=2.3)

    with pytest.raises(ValueError):
        big_int = 10
        close_up(df, n=big_int)


tests/test_pyeasyeda.py
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pytest
from pytest import raises
from pyeasyeda import pyeasyeda as pye
    
def test_birds_eye_view_error():
    """tests birds_eye_view handling of erroneous inputs"""
    
    #Load toy data set
    data = {'product_name': 
            ['laptop', 'printer', 'tablet', 'laptop', 'printer', 'laptop'],
            'price': [1200, 150, 300, 450, 200, 1000],
            'weight':[200, 500, 700, 100, 1000, 800]}
    
    eda = pd.DataFrame(data)
    
    #Section of invalid inputs
    with raises(TypeError):
        pye.birds_eye_view(1)

    with raises(TypeError):
        pye.birds_eye_view(eda, var_list="list")

    with raises(TypeError):
        pye.birds_eye_view(eda, n="2")
    
    with raises(TypeError):
        pye.birds_eye_view(eda, var_list=['notreal'])
    
def test_birds_eye_view_outputs():
    """tests birds_eye_view vizualization outputs"""
    
    #Load toy data set
    data = {'product_name': 
            ['laptop', 'printer', 'tablet', 'laptop', 'printer', 'laptop'],
            'price': [1200, 150, 300, 450, 200, 1000],
            'weight':[200, 500, 700, 100, 1000, 800],
            'version':[400, 600, 200, 700, 2000, 100]}
    
    eda = pd.DataFrame(data)
    
    viz = pye.birds_eye_view(eda)
    viz_custom = pye.birds_eye_view(eda, var_list=['price','weight','product_name'])
    
    # Test birds_eye_view with default inputs
    
    # Test if each different vizualization type generated, that the correct amount generated, and that it saved correctly - default
    assert (
        type(viz["histograms"]) == list
    ), "Histogram objects should be stored in list of figures"
    assert (
        type(viz["bar_charts"]) == list
    ), "Bar Chart objects should be stored in list figures"
    assert (
            str(type(viz["heatmap"])) == "<class 'matplotlib.axes._subplots.AxesSubplot'>"

    ), "Heatmap should be stored as a single matplotlib figure"
    
    assert (
        len(viz) == 3
    ), "Number of items in the viz dictionary should be 3" 
    assert (
        len(viz["histograms"]) == 3
    ), "Number of items in the 'histograms' key should be 3"
    assert (
        len(viz["bar_charts"]) == 1
    ), "Number of items in the 'bar_charts' key should be 1" 
    assert (
        "heatmap" in viz.keys()
    ), "There should be a 'heatmap' key in the viz dictionary"

    # Test if each different vizualization type generated, that the correct amount generated, and that it saved correctly - custom
    assert (
        type(viz_custom["histograms"]) == list
    ), "Histogram objects should be stored in list of figures"
    assert (
        type(viz_custom["bar_charts"]) == list
    ), "Bar Chart objects should be stored in list figures"
    assert (
            str(type(viz_custom["heatmap"])) == "<class 'matplotlib.axes._subplots.AxesSubplot'>"

    ), "Heatmap should be stored as a single matplotlib figure"
    
    assert (
        len(viz_custom) == 3
    ), "Number of items in the viz dictionary should be 3" 
    assert (
        len(viz_custom["histograms"]) == 2
    ), "Number of items in the 'histograms' key should be 2"
    assert (
        len(viz_custom["bar_charts"]) == 1
    ), "Number of items in the 'bar_charts' key should be 1" 
    assert (
        "heatmap" in viz_custom.keys()
    ), "There should be a 'heatmap' key in the viz dictionary"

    

src/pyeasyeda/close_up.py
import pandas as pd
import numpy as np
import altair as alt


def close_up(df, n=1):
    """Takes in a dataframe object and the number of pairs of strongest correlations, and
    returns scatterplots with a correlation trend for each pair.

        Parameters
        ----------
        df : pd.core.frame.DataFrame
             dataframe to create the visualization(s)
        n : int
            number of visualizations of the variable(s) with the 
            strongest correlation to the dependent variable to be displayed,
            defaults to 1

        Examples
        --------
        >>> close_up(df, n = 4)
    """

    # check if input is a DataFrame
    if not isinstance(df, pd.core.frame.DataFrame):
        raise TypeError("df should be of type 'pandas.core.frame.DataFrame'")

    if not isinstance(n, int):
        raise TypeError("n should be of type 'int'.")

    # calculate max allowable integer
    numeric = df.select_dtypes(include=np.number).columns.tolist()
    corr_matrix = df[numeric].corr()
    N_max = len(corr_matrix) * (len(corr_matrix) - 1) / 2

    # check if input exceeds max allowable integer
    if n > N_max:
        raise ValueError("n exceeds total number of coefficients.")

    # initialization
    corr = corr_matrix.to_numpy()
    corr = abs(corr)
    corr = np.triu(corr)  # take only upper triangle
    np.fill_diagonal(corr, 0)  # zero diagonal
    max_row = np.argmax(np.max(corr, axis=1))
    max_col = np.argmax(np.max(corr, axis=0))
    viz = {}  # viz dict to be returned

    # plot
    for i in range(n):
        max_row = np.argmax(np.max(corr, axis=1))
        max_col = np.argmax(np.max(corr, axis=0))
        coef = corr_matrix.iloc[max_row, max_col]
        points = (
            alt.Chart(df, title=f'coeff: {coef:.3f}')
            .mark_point(opacity=0.3)
            .encode(
                alt.X(corr_matrix.columns[max_row]), alt.Y(
                    corr_matrix.columns[max_col])
            )
        )
        viz[i+1] = points + points.transform_regression(
            corr_matrix.columns[max_row], corr_matrix.columns[max_col]).mark_line(size=3)
        corr[max_row, max_col] = 0

    # generate one big chart
    chart = viz[1]
    for i in range(n-1):
        chart &= viz[i+2]

    return chart


src/pyeasyeda/birds_eye_view.py
import numpy as np
import pandas as pd
import seaborn as sns
import warnings
import matplotlib.pyplot as plt

def birds_eye_view(df, n=20, var_list=None):
    """Takes in a pandas.DataFrame object, an optional integer for the histogram bin size, an optional custom variable list, and displays 3 different visualization sets.

    1. Histograms for each numeric variable
    2. A bar chart for each categorical variable
    3. A correlation heatmap of the numeric variables.

    Parameters
    ----------
    df : pandas.DataFrame
        dataframe to create the visualizations
    n : int
        bin size for histograms
    var_list : list
        a specific list of variables to examine, defaults to None

    Returns
    -------
    charts: dict
        A dictionary containing the plot objects created by this function

    Examples
    --------
    >>> birds_eye_view(df, n=30)

    """

    if not (type(df) == pd.DataFrame):
        raise TypeError("df must be input as a DataFrame.")

    if type(var_list) != list and var_list is not None:
        raise TypeError("var_list must be a list.")

    if type(n) != int:
        raise TypeError("n must be an integer.")

    # Generate the visualizations

    viz = {}
    heatmap_list = []
    heatmap = []
    histograms = []
    bar_charts = []

    # Defining the numeric and categorical variables
    numeric = df.select_dtypes(include=np.number).columns.tolist()
    categorical = df.select_dtypes(include=("object" or "string")).columns.to_list()

    # Plot all the variables
    if var_list is None:

        # Histograms
        for num_col in numeric:
            chart = sns.histplot(df, x=(num_col), bins=n, kde=True)
            plt.title("Histogram for " + num_col)
            plt.figure()
            histograms.append(chart)

        # Bar Charts
        for cat_col in categorical:
            if len(pd.unique(df[cat_col])) > 11:
                print(cat_col, " has too many unique values")
            else:
                chart = sns.countplot(data=df, x=(cat_col))
                plt.title("Bar Chart for " + cat_col)
                plt.figure()
                bar_charts.append(chart)

        # Heatmap
        corr_matrix = df[numeric].corr()
        mask = np.triu(np.ones_like(corr_matrix, dtype=np.bool_))
        chart = sns.heatmap(data=corr_matrix,
                            vmin=-1,
                            vmax=1,
                            annot=True,
                            cmap="BrBG",
                            mask=mask
        )
        plt.title("Heatmap of correlation between numeric features")
        plt.figure(figsize=(12, 6))
        print(chart)
        viz["heatmap"] = chart

    # Plot just the custom variables from var_list (if applicable)
    else:
        
        for custom_col in var_list:
            all_cols = df.columns.to_list()
            if custom_col not in all_cols:
                raise TypeError("Variable name " +
                                custom_col +
                                " not found in data frame, please check inputs in var_list.")

            # Histograms
            if custom_col in numeric:
                heatmap_list.append(custom_col)
                chart = sns.histplot(df, x=(custom_col), bins=n, kde=True)
                plt.title("Histogram for " + custom_col)
                plt.figure()
                histograms.append(chart)

            # Bar Charts
            elif custom_col in categorical:
                if len(pd.unique(df[custom_col])) > 11:
                    print(cat_col, " has too many unique values")
                else:
                    chart = sns.countplot(data=df, x=(custom_col))
                    plt.title("Bar Chart for " + custom_col)
                    plt.figure()
                    bar_charts.append(chart)

        # Heatmap
        corr_matrix = df[heatmap_list].corr()
        mask = np.triu(np.ones_like(corr_matrix, dtype=np.bool_))
        chart = sns.heatmap(data=corr_matrix,
                            vmin=-1,
                            vmax=1,
                            annot=True,
                            cmap="BrBG",
                            mask=mask
        )
        plt.title("Heatmap of correlation between numeric features")
        plt.figure(figsize=(12, 6))
        viz["heatmap"] = chart

    viz["histograms"] = histograms
    viz["bar_charts"] = bar_charts
    
    return viz

src/pyeasyeda/summary_suggestions.py
import pandas as pd
import numpy as np

def summary_suggestions(df, threshold = 0.8):
    """Takes a dataframe object and displays a table of summary statistics. 
    Underneath that table it prints analysis considerations that 
    help highlight potential issues, if applicable, and serves as a general 
    guideline for the analysis.

    Parameters
    ----------
    df : pandas dataframe
        Dataframe to be examined

    threshold : float
        threshold for considering class imbalance

    Returns
    -------

    numeric_summary_df : pandas dataframe
        Dataframe for summary statistics of numeric variables

    categorical_summary_df : pandas dataframe
        Dataframe for summary statistics of categorical variables

    Examples
    --------
    >>> summary_suggestions(df)

    (summary statistics for numeric variables)
        
    ** Potential Analysis Considerations: **
    Please review and take the following into consideration:
    
    Class Imbalance: There appears to be class imbalance in the variable X. 
    Unique Values: Variable X is comprised of 97% unique values in it's observations.'
    
    """

    # check if input is a DataFrame
    if not isinstance(df, pd.DataFrame):
        raise TypeError("Input df must be a pandas dataframe object")

    if type(threshold) != int:
        raise TypeError("Input threshold must be an integer")

    if not isinstance(numeric_summary_df, pd.DataFrame):
        raise TypeError("Output summary dataframe df for numeric columns must be a pandas dataframe object")

    if not isinstance(categorical_summary_df, pd.DataFrame):
        raise TypeError("Output summary dataframe df for categorical columns must be a pandas dataframe object")

    numeric_summary_df = df.select_dtypes(include=np.number).describe()
    categorical_summary_df = df.select_dtypes(include=np.object_).describe()

    print("** Potential Analysis Considerations: ** \n \
    Please review and take the following into consideration:")

    print(f"Summary statistics for numeric columns of the dataframe: \n\n {numeric_summary_df} \n\n")
    print(f"Summary statistics for categorical columns of the dataframe: \n\n {categorical_summary_df} \n\n")

    class_imbalance_df = categorical_summary_df[categorical_summary_df.index == 'unique']/len(df)
    filtered_class_imbalance_df = class_imbalance_df.loc['unique'] > threshold
    class_imbalance_vars = [*filter(class_imbalance_df.get, class_imbalance_df.index)]
    
    for var in class_imbalance_vars:
        print(f"There appears to be class imbalance in variable {var} \
        which comprises of {round(class_imbalance_df[var]['unique']*100)}% unique values. \n ")



src/pyeasyeda/pyeasyeda.py
import numpy as np
import pandas as pd
import seaborn as sns
import warnings
import matplotlib.pyplot as plt

def clean_up(df):
    """Takes a dataframe object and returns a cleaned version with the NaN values removed.
     It also prints a list of potential outliers for each explanatory variable, based on a
     threshold distance from the respective median value.
        Parameters
        ----------
        df : dataframe
            dataframe to be cleaned
    
        Returns
        -------
        df_clean
            same dataframe with all the NaN's removed
        Examples
        --------
        >>> df_clean = clean_up(df)
                
        '**The following potenital outliers were detected:**
        Variable X: 100002039.26, located at [0,4]
        Variable Y: .2, located at [3,56]'
    
    """
    return 

def birds_eye_view(df, n=20, var_list=None):
    """Takes in a pandas.DataFrame object, an optional integer for the histogram bin size, an optional custom variable list, and displays 3 different visualization sets.

    1. Histograms for each numeric variable
    2. A bar chart for each categorical variable
    3. A correlation heatmap of the numeric variables.

    Parameters
    ----------
    df : pandas.DataFrame
        dataframe to create the visualizations
    n : int
        bin size for histograms
    var_list : list
        a specific list of variables to examine, defaults to None

    Returns
    -------
    charts: dict
        A dictionary containing the plot objects created by this function

    Examples
    --------
    >>> birds_eye_view(df, n=30)

    """

    if not (type(df) == pd.DataFrame):
        raise TypeError("df must be input as a DataFrame.")

    if type(var_list) != list and var_list is not None:
        raise TypeError("var_list must be a list.")

    if type(n) != int:
        raise TypeError("n must be an integer.")

    # Generate the visualizations

    viz = {}
    heatmap_list = []
    heatmap = []
    histograms = []
    bar_charts = []

    # Defining the numeric and categorical variables
    numeric = df.select_dtypes(include=np.number).columns.tolist()
    categorical = df.select_dtypes(include=("object" or "string")).columns.to_list()

    # Plot all the variables
    if var_list is None:

        # Histograms
        for num_col in numeric:
            chart = sns.histplot(df, x=(num_col), bins=n, kde=True)
            plt.title("Histogram for " + num_col)
            plt.figure()
            histograms.append(chart)

        # Bar Charts
        for cat_col in categorical:
            if len(df[cat_col].value_counts()) > 10:
                message = (
                    "The variable "
                    + cat_col
                    + " has more than 10 "
                    + "unique values, and is not suitable for a bar chart."
                )
                warnings.warn(message, DeprecationWarning)
            else:
                chart = sns.countplot(data=df, x=(cat_col))
                plt.title("Bar Chart for " + cat_col)
                plt.figure()
                bar_charts.append(chart)

        # Heatmap
        corr_matrix = df[numeric].corr()
        mask = np.triu(np.ones_like(corr_matrix, dtype=np.bool_))
        chart = sns.heatmap(data=corr_matrix,
                            vmin=-1,
                            vmax=1,
                            annot=True,
                            cmap="BrBG",
                            mask=mask
        )
        plt.title("Heatmap of correlation between numeric features")
        plt.figure(figsize=(12, 6))
        print(chart)
        viz["heatmap"] = chart

    # Plot just the custom variables from var_list (if applicable)
    else:
        
        for custom_col in var_list:
            all_cols = df.columns.to_list()
            if custom_col not in all_cols:
                raise TypeError("Variable name " +
                                custom_col +
                                " not found in data frame, please check inputs in var_list.")

            # Histograms
            if custom_col in numeric:
                heatmap_list.append(custom_col)
                chart = sns.histplot(df, x=(custom_col), bins=n, kde=True)
                plt.title("Histogram for " + custom_col)
                plt.figure()
                histograms.append(chart)

            # Bar Charts
            elif custom_col in categorical:
                if len(df[custom_col].value_counts()) > 10:
                    message = (
                        "The variable "
                        + custom_col
                        + " has more than 10 "
                        + "unique values, and is not suitable for a bar chart."
                    )
                    warnings.warn(message, DeprecationWarning)
                else:
                    chart = sns.countplot(data=df, x=(custom_col))
                    plt.title("Bar Chart for " + custom_col)
                    plt.figure()
                    bar_charts.append(chart)

        # Heatmap
        corr_matrix = df[heatmap_list].corr()
        mask = np.triu(np.ones_like(corr_matrix, dtype=np.bool_))
        chart = sns.heatmap(data=corr_matrix,
                            vmin=-1,
                            vmax=1,
                            annot=True,
                            cmap="BrBG",
                            mask=mask
        )
        plt.title("Heatmap of correlation between numeric features")
        plt.figure(figsize=(12, 6))
        viz["heatmap"] = chart

    viz["histograms"] = histograms
    viz["bar_charts"] = bar_charts
    
    return viz

def close_up(df, N=1):
    """Takes in a dataframe object and displays a scatterplot with a correlation trend 
    line visualization of the variable(s) most strongly correlated with the 
    dependent variable.
        Parameters
        ----------
        df : dataframe
            dataframe to create the visualization(s)
        N : number of visualizations of the variable(s) with the 
            strongest correlation to the dependent variable to be displayed,
            defaults to 1
        
        Examples
        --------
        >>> close_up(df_clean)
        
        (displays visualizations)
    
    """
    return

def summary_suggestions(df):
    """Takes a dataframe object and displays a table of summary statistics. 
    Underneath that table it prints analysis considerations that 
    help highlight potential issues, if applicable, and serves as a general 
    guideline for the analysis.
        Parameters
        ----------
        df : dataframe
            Dataframe to be examined
        
        Examples
        --------
        >>> summary_suggestions(df_clean)
        
        '(summary statistics)
        
        **Potential Analysis Considerations:**
        Please review, and take the following into consideration:
        
        Class Imbalance: There appears to be class imbalance in the variable X. 
        Unique Values: Variable X is comprised of 97% unique values in it's observations.'
    
    """
    return
